pandas/tseries/holiday.py

```diff
@@ -152,14 +159,14 @@ class Holiday:
     def __init__(
         self,
         name,
-        year=None,
-        month=None,
-        day=None,
-        offset=None,
+        year: int | None = None,
+        month: int = 1,
+        day: int = 1,
```

Comment:
`Holiday` is only ever called with `month` and `day`. Could also make them keyword-only in case the default value is not wanted.

Reply:
I tend to prefer PRs adding type annotations not to change things and if a issue is found, opening an issue to discuss. Is the Holiday class public? If not, I guess no harm. 

> `Holiday` is only ever called with `month` and `day`. Could also make them keyword-only in case the default value is not wanted.

I assume you meant positional without defaults?

Reply:
> I tend to prefer PRs adding type annotations not to change things

Could change it to `int | None`, add a deprecation/future warning, and add mypy ignores. `day` and `month` are forwarded to `datetime` which requires an `int`. Mypy currently does not find this issue as it infers them to be of type `Any` (pyright infers `Unknown | None`).

> Is the Holiday class public? 

Based on the doc-definition of public it [doesn't seem so](https://pandas.pydata.org/pandas-docs/stable/search.html?q=Holiday), only concrete instances of `Holiday` such as `Easter`.

> I assume you meant positional without defaults?

I meant required keyword-only arguments

```py
def __init__(
        self,
        name,
        year: int | None=None,
        *,
        month: int,  # required keyword-only argument
        day: int, # required keyword-only argument
        ...) -> None: ...
```

Reply:
I think `Holiday` is technically public as it's in `pandas.tseries` and "mentioned in the documentation": https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#holidays-holiday-calendars

Reply:
Since
```py
import pandas as pd

pd.tseries.holiday.Holiday("test").dates(pd.Timestamp.now(), pd.Timestamp.now())
...
    datetime(start_date.year - 1, self.month, self.day)
TypeError: 'NoneType' object cannot be interpreted as an integer
```

errors, would it be okay to raise a TypeError already in `__init__`?

Reply:
Though a user I guess could update the attributes by setting them after they created the instance, I would be +1 just to raise a `TypeError` in `__init__`. It doesn't appear that this class is very useful if some of these keywords are not specified.

Reply:
I will close this for now and later revisit raising a `TypeError` inside `__init__`.

Reply:
Sounds good. Would be good for `Holiday` to be added to the API reference documentation as well